working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+> 64-bits wide. We might want to eschew atomicity guarantees and allow load
+> tearing for primitive fields - though it makes semantics somewhat
+> unpredictable and architecture dependent. The same concern applies to unboxed
+> SIMD values inside `Shareable` objects.
+
+### Shareable Types
+
+> Type is _shareable_ if and only if one of the following applies:
+>
+> - It is `Null` or `Never`.
+> - It is an interface type which is subtype of `Shareable`.
+> - It is a record type where all field types are shareable.
+> - It is a nullable type `T?` where `T` is shareable type.
+
+### Why type based opt-in?
+
+There are two main reasons for choosing explicit opt-in into shareability:
+
+1. **Aligning with emerging JS and Wasm capabilities.** Current proposals for
+   shared memory multithreading on the Web (see details
+   [below](#web-js-and-wasm)) propose partitioning shareable objects from
+   non-shareable. We would like to make sure that Dart's semantics is possible
+   to translate to both JS and Wasm - allowing us to fully implement our
+   concurrency story on the Web.
+2. **Enabling graceful adoption of shared memory multithreading.** Shared memory
+   multithreading is complicated enough by itself, so I feel that it would
+   complicate things more if we were to forcefully bring all existing libraries
+   (not written with shared memory in mind) into the world where any data can be
+   accessed from multiple threads at once. Existing code will continue to run
+   _as is_ within isolates. Newly written code can choose to opt-in into
+   shareability where it matters for performance or interoperability reasons.
+   Marking class as `Shareable` gives a strong signal that the developer has
+   considered implications of sharing instances of these class across threads
+   and took measure to ensure that it is safe.
+
+I would like to introduce shared memory multithreading into Dart in a way that
+avoids subtle breakages in the existing code. Consider for a moment a library
+which maintains a global cache internally and is written under the assumption
+that Dart is a single threaded:
+
+```dart
+class C {
+  final id;
+  C(this.id);
+}
+
+int _nextId = 0;
+final _cache = <int, C>{};
+
+C makeObject() {
+  _cache[nextId] = C(nextId);
+  nextId++;
+  return _cache[nextId - 1];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the
+same code becomes thread _unsafe_ in the presence of shared memory
+multithreading.
+
+### Generics
+
+When declaring a generic type the developer will have to use type parameter
+bounds to ensure that resulting class conforms to restrictions imposed by
+`Shareable`:
+
+```dart
+class X<T> implements Shareable {
+  T v;  // compile time error.
+}
+
+class Y<T extends Shareable> implements Shareable {
+  T v;  // ok.
+}
+```
+
+> [!NOTE]
+>
+> We could really benefit from the ability to use intersection types here, which
```

Comment:
Link to https://github.com/dart-lang/language/issues/2709 and/or https://github.com/dart-lang/language/issues/1152.